## 7장 : 오류 처리

- 상당수의 코드 기반은 전적으로 오류 처리 코드에 좌우된다.
    - 여기저기 흩어진 오류 처리 코드 때문에 실제 코드가 하는 일을 파악하기가 거의 불가능하다.
    - 오류 처리 코드로 인해 프로그램 논리를 이해하기 어려워진다면 깨끗한 코드라 부르기 어렵다.
    - 오류 처리는 중요하다. 하지만 로직을 헷갈리게 만드는 오류처리는 나쁘다.

### 오류 코드보다 예외를 사용하라

- Error Handling
    - Error code
    - Exception

[Yesterday I was having a heated debate with a coworker](https://stackoverflow.com/questions/253314/conventions-for-exceptions-or-error-codes)

- 과거에는 예외를 지원하지 않는 프로그래밍 언어가 대부분이었고, 이들은 에러를 처리하기 위해 오류 코드를 사용했었다.
- 하지만 최근 언어들은 예외 처리를 지원하고, 이 방법이 더욱 깔끔하고 간결하다.

```java
// Bad
public class DeviceController {
  ...
  public void sendShutDown() {
    DeviceHandle handle = getHandle(DEV1);
    // Check the state of the device
    if (handle != DeviceHandle.INVALID) {
      // Save the device status to the record field
      retrieveDeviceRecord(handle);
      // If not suspended, shut down
      if (record.getStatus() != DEVICE_SUSPENDED) {
        pauseDevice(handle);
        clearDeviceWorkQueue(handle);
        closeDevice(handle);
      } else {
        logger.log("Device suspended. Unable to shut down");
      }
    } else {
      logger.log("Invalid handle for: " + DEV1.toString());
    }
  }
  ...
}


// Good
public class DeviceController {
  ...
  public void sendShutDown() {
    try {
      tryToShutDown();
    } catch (DeviceShutDownError e) {
      logger.log(e);
    }
  }
    
  private void tryToShutDown() throws DeviceShutDownError {
    DeviceHandle handle = getHandle(DEV1);
    DeviceRecord record = retrieveDeviceRecord(handle);
    pauseDevice(handle); 
    clearDeviceWorkQueue(handle); 
    closeDevice(handle);
  }
  
  private DeviceHandle getHandle(DeviceID id) {
    ...
    throw new DeviceShutDownError("Invalid handle for: " + id.toString());
    ...
  }
  ...
}
```

#### 추가

위의 전제

1. 부수 효과가 없다면
1. 에러가 발생했을 때 별 다른 조치를 안해도 된다면

오류를 처리하는 방법들

1. 오류 코드를 사용

```go
var ERROR = 10000000000

func div(num1 float32, num2 float32) float32 {
    if (num2 == 0) {
        return float32(ERROR)
    } else {
        return num1 / num2
    }
}

var result float32 = div(num1, num2)

if int(result) == ERROR {
    fmt.Println("에러!!")
} else {
    fmt.Println(result)
}
```
-> 에러처리 구문인지 분기문인지 헷갈린다

2. 예외 처리

```java

public class Main {
    public static void main(String[] args) {
        int num1 = 10;
        int num2 = 0;
        try{
            System.out.println(num1 / num2);
        }
        catch (ArithmeticException e){
            System.out.println("에러!!");
        }
    }
}
```

3. 에러를 함께 반환

```go
import "errors"

var errorDivByZero = errors.New("You can't divide by zero.")

func div(num1 float32, num2 float32) (float32, error) {
    if (num2 == 0) {
        return 0, errorDivByZero
    } else {
        return num1 / num2, nil
    }
}

func main() {
    var num1 float32 = 3
    var num2 float32 = 2
    result, err := div(num1, num2)
    if err != nil {
        fmt.Println(err)
        return
    } 
    fmt.Println(result)
}
```

### Try - Catch - Finally 문부터 작성하라

- **예외가 발생하는 코드를 짤 때는** try-catch-finally 문으로 시작하는 편이 낫다.

### Unchecked Exceptions 를 사용하라

Checked, Unchecked Exceptions -> Java 에서 사용되는 Exception 개념 중 하나.

Checked Exception 이 발생할 수 있는 메소드를 사용할 경우 반드시 예외를 처리하는 코드를 함께 작성해야 한다. 그렇지 않으면 컴파일 에러.

Unchecked Exceptions 는 예외 처리를 하지 않아도 컴파일이 가능하기 때문에 Unchecked 로 이름 붙여졌다.

[참고 1](https://hyeonstorage.tistory.com/199)
[참고 2](http://www.nextree.co.kr/p3239/)

```
특정 메소드에서 checked exception을 throw하고
3단계(메소드 콜) 위의 메소드에서 그 exception을 catch한다면
모든 중간단계 메소드에 exception을 정의해야 한다.(자바의 경우 메소드 선언에 throws 구문을 붙이는 등)
상위 레벨 메소드에서 하위 레벨 메소드의 디테일에 대해 알아야 하기 때문에 캡슐화 또한 깨진다.
필요한 경우 checked exceptions를 사용해야 되지만 일반적인 경우 득보다 실이 많다.
```

한마디로, 모호한 Exception 을 사용하는게 여러모로 편하다

### 정상 흐름을 정의하라

```python
# 1
students = count_students(people)
if students is not None:
    for student in students:
        # ...

# 2
students = count_students(people)
for student in students:
    # ...


def count_students(people):
    students = []
    for person in people:
        if person.job == 'STUDENT':
            students.append(person)
    return students
```

- 클래스나 객체가 예외적인 상황을 캡슐화해서 처리하자
- 코드를 부르는 입장에서(클라이언트 코드) 예외적인 상황을 신경쓰지 않아도 된다.

```java
// 호출하는 쪽에서 매번 따로 확인하는 것은 번거롭다.
try {
    MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
    m_total += expenses.getTotal();
} catch(MealExpensesNotFound e) {
    m_total += getMealPerDiem(); // 호출하는 쪽에서 매번 따로 확인하는 것은 번거롭다.
}


// 호출하는 쪽은 함수를 부르기만 하고
MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
m_total += expenses.getTotal();

public class ExpenseReportDAO {
    ...
    public MealExpenses getMeals(int employeeId) {
        MealExpenses expenses;
        // 클래스 내부에서 특이 케이스를 처리한다
        try {
            expenses = expenseReportDAO.getMeals(employee.getID());
        } catch(MealExpensesNotFound e) {
            expenses = new PerDiemMealExpenses();
        }
      
        return expenses;
    }
    ...
}
```

### null 을 전달하지 마라

- Method 에서 null 을 반환하는 방식도 나쁘지만 method 로 null 을 전달하는 방식은 더 나쁘다.

```java
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    return (p2.x – p1.x) * 1.5;
  }
  ...
}


public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    if(p1 == null || p2 == null){
      throw InvalidArgumentException("Invalid argument for MetricsCalculator.xProjection");
    }
    return (p2.x – p1.x) * 1.5;
  }
}
```

> 어느 단계까지 의심할 것이며 어느 단계부터 안전성을 보장할 것인가??

### 추가

- 에러를 처리할 때 고려해야 할 점
    1. 죽일 것 인가? 살릴 것 인가? 
    1. 가까운 곳에서 즉시 처리할 것인가? 상위로 위임할 것인가?
    1. 강제적으로 처리 시킬 것인가? 아닌가?
    1. 모두 되돌릴 것인가? 이미 벌어진 일은 무시, 포기할 것인가?
- 결과적으로
    - Exception 이나 Error code 목표는 같다. 잘 쓰면 서로 비슷하게 동작할 수 있음.
    - 다만 Exception 은 개발자를 유혹에 빠뜨린다. 결국 죽여야하는 타이밍을 놓치게 된다.

```python
def my_function():
    try:
        start()
        process()
        another_process()
        post_process()
        last_process()
    except Exception as e:
        print(e)
        return False
    return True

```

## 8장 : 경계

1. 내가 구현한 코드와 외부 코드 사이의 **경계**
1. 내가 아는 코드와 모르는 코드 사이의 **경계**
1. 이미 작성된 코드와 아직 작성하지 않은 코드 사이의 **경계**

### 경계 살피고 익히기

- 외부 패키지 테스트가 우리의 책임은 아니지만, 우리 자신을 위해 우리가 사용할 코드를 테스트하는 편이 바람직하다
- 학습 테스트 :  외부 코드에 대한 테스트 코드를 작성하며 그 코드를 더 자세히 학습하는 것

### 학습 테스트는 공짜 이상이다

- 서드파티 패키지에 새로운 업데이트가 있다면 학습 테스트를 돌려 차이가 있는지 확인할 수 있다.
    - 이런 경계 테스트가 없다면 낡은 버전을 필요 이상으로 오랫동안 사용하려는 유혹에 빠지기 쉽다
    - 

### 아직 존재하지 않는 코드 사용하기

Mocking, Dummy API 등을 사용해 임시로 구현하고 향후 대체하기

### 깨끗한 경계

- 통제하지 못하는 코드를 사용할 때는 각별한 주의가 필요하다.
- 경계에 위치하는 코드는 깔끔히 분리한다.
- **외부 패키지를 호출하는 코드를 가능한 줄여 경계를 관리하자.**
    - 가독성이 높아지고 일관성도 높아지며 외부 패키지가 변했을 때 변경할 코드도 줄어든다.
